flocker/node/_deploy.py

```diff
@@ -424,6 +453,34 @@ def calculate_necessary_state_changes(self, desired_state,
         :return: A ``Deferred`` which fires with a ``IStateChange``
             provider.
         """
+        # Add missing dataset IDs to the desired configuration. Should be
+        # done elsewhere: https://clusterhq.atlassian.net/browse/FLOC-1199
+
+        # We don't want to mutate the desired configuration, so do a
+        # deepcopy (PMap doesn't support copy.deepcopy):
+        desired_state = loads(dumps(desired_state))
+
+        datasets_with_no_id = []
+        for node in desired_state.nodes:
+            for application in node.applications:
+                if application.volume:
+                    dataset = application.volume.dataset
+                    if dataset.dataset_id is None:
+                        datasets_with_no_id.append(dataset)
+        current_datasets_by_name = {}
+        for node in current_cluster_state.nodes:
+            for application in node.applications:
+                if application.volume:
+                    dataset = application.volume.dataset
+                    name = dataset.metadata[u"name"]
+                    current_datasets_by_name[name] = dataset
+        for dataset in datasets_with_no_id:
+            matching = current_datasets_by_name.get(dataset.metadata[u"name"])
+            if matching:
+                dataset.dataset_id = matching.dataset_id
+            else:
+                dataset.dataset_id = unicode(uuid4())
+
```

Comment:
This could probably be refactored.  A method to compute `datasets_with_no_id`.  A method to compute `current_datasets_by_name`.  A method to copy `desired_state`, mutate the copy given the previous two pieces of information, and return the copy.  This at least keeps the growth of this already too-long method to a minimum.